assert: improve error messages

BridgeAR opened the pull request:
This switches all assert error messages to use the new diffing output as in the `assert` strict mode.

It also makes sure not reference equal objects are now visualised as such. Before it would e.g. have been `[1, 2] strictEqual [1, 2]` and that would be quite confusing. That is now detected and [expressed](https://github.com/nodejs/node/compare/master...BridgeAR:improve-error-messages?expand=1#diff-81c8c31c5728ec9ed47700e55d9ceebfR870).

I also added support in util.inspect to visualize arguments to distinguish them from a regular object.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13754/

BridgeAR commented:
I am uncertain why the CI fails. Could someone please take a look?

BridgeAR commented:
CITGM PR https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1338/
CITGM master https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1339/

BridgeAR commented:
Seems like this breaks `eslint` tests. @not-an-aardvark can you PTAL?

It does not seem like a big problem to me but I might be wrong.

Here is a small patch that might fix it (without verifying it for real):

```diff
@@ -439,7 +439,7 @@ class RuleTester {
                     `Expected '${actual}' to match ${expected}`
                 );
             } else {
-                assert.strictEqual(actual, expected);
+                assert.strictEqual(actual, expected, util.format("%s strictEqual %s", actual, expected));
             }
         }
```

Looking at it further, it might make sense to change some more assert things as well like e.g.:

```diff
diff --git a/lib/testers/rule-tester.js b/lib/testers/rule-tester.js
index de218a875..798253368 100644
--- a/lib/testers/rule-tester.js
+++ b/lib/testers/rule-tester.js
@@ -396,11 +396,11 @@ class RuleTester {
          * @private
          */
         function assertASTDidntChange(beforeAST, afterAST) {
-            if (!lodash.isEqual(beforeAST, afterAST)) {
-
-                // Not using directly to avoid performance problem in node 6.1.0. See #6111
-                // eslint-disable-next-line no-restricted-properties
-                assert.deepEqual(beforeAST, afterAST, "Rule should not modify AST.");
+            if (util.isDeepStrictEqual) {
+                assert.deepStrictEqual(beforeAST, afterAST, "Rule should not modify AST.");
+                // Not using directly to avoid performance problem in Node.js 6 and 7. See #6111
+            } else if (!lodash.isEqual(beforeAST, afterAST)) {
+                assert.fail("Rule should not modify AST.");
             }
         }
 
@@ -539,7 +539,7 @@ class RuleTester {
                     } else {
 
                         // Message was an unexpected type
-                        assert.fail(message, null, "Error should be a string, object, or RegExp.");
+                        assert.fail("Error should be a string, object, or RegExp.");
                     }
                 }
             }
@@ -554,8 +554,7 @@ class RuleTester {
                 } else {
                     const fixResult = SourceCodeFixer.applyFixes(item.code, messages);
 
-                    // eslint-disable-next-line no-restricted-properties
-                    assert.equal(fixResult.output, item.output, "Output is incorrect.");
+                    assert.strictEqual(fixResult.output, item.output, "Output is incorrect.");
                 }
             }
 ```

not-an-aardvark commented:
I agree that the `eslint` test failure is not a big problem (it looks like no significant functionality is broken).

Presumably, `eslint` users would also benefit from improved assertion error messages, so I'm not sure it would be a good idea to lock the error messages within `eslint` just to get the existing tests to pass. I suppose we could do feature-detection in the tests to check what error message to expect, e.g.:

```js
assert.strictEqual(message, new assert.AssertionError({ actual: 5, expected: 'foo', operator: '===' }).message);
```

Trott commented:
If we do feature detection, maybe include a comment about when we can drop the feature detection. (In other words, "this feature detection code won't be needed anymore after we drop support for Node.js 8.0.0" or whatever.)

BridgeAR commented:
Rebased due to conflicts.

New CI https://ci.nodejs.org/job/node-test-pull-request/13857/

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14004/

BridgeAR commented:
@jasnell @mcollina PTAL

jasnell commented:
Still lgtm

Trott commented on lib/internal/errors.js:
```diff
@@ -19,6 +19,13 @@ let green = '';
 let white = '';
 
+const READABLE_OPERATOR = {
+  deepStrictEqual: 'Input A expected to strictly deep equal input B',
```
maybe `strictly deep-equal`?

Trott commented on lib/internal/errors.js:
```diff
@@ -19,6 +19,13 @@ let green = '';
 
+const READABLE_OPERATOR = {
+  deepStrictEqual: 'Input A expected to strictly deep equal input B',
+  notDeepStrictEqual: 'Input A expected to strictly not deep equal input B',
```
Same here? `deep equal` -> `deep-equal`?

Trott commented:
Not sure how I feel about `Input A` and `Input B` being used for "first argument" and "second argument". Maybe `Argument 1` and `Argument 2` would be less problematic?

BridgeAR commented:
I addressed the comment. @Trott I would like to keep `Input A / B` for now as that could be changed in a different PR. Are you OK with that, so this can land?

Trott commented:
> I addressed the comment. @Trott I would like to keep Input A / B for now as that could be changed in a different PR. Are you OK with that, so this can land?

Yes, that's fine by me.

BridgeAR commented:
It would be nice to get another LG before landing.

BridgeAR commented on test/message/assert_throws_stack.out:
```diff
@@ -2,7 +2,7 @@ assert.js:*
   ^
 
-AssertionError [ERR_ASSERTION]: Input A expected to deepStrictEqual input B:
+AssertionError [ERR_ASSERTION]: Input A expected to strictly deep equal input B:
```
Thanks a lot!

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/14057/

BridgeAR commented:
@nodejs/testing @nodejs/tsc PTAL. I would like to land this soon.

Trott commented:
Since it's semver-major, it needs a CITGM run, if I'm not mistaken.

BridgeAR commented:
@Trott I did that already in the beginning. It breaks the eslint tests and I opened a PR in eslint to address that. Besides that there was nothing and that should not be blocking.

Trott commented:
> @Trott I did that already in the beginning.

Not sure how I missed that. Sorry!

BridgeAR commented:
If there is no objection until then, I am going to land this in ~30 hours.

targos commented:
No objections from me

Trott commented:
Re-running failed Windows CI: https://ci.nodejs.org/job/node-test-commit-windows-fanned/17190/

BridgeAR commented:
Rebased due to minor conflicts in the tests.

One more CI before landing: https://ci.nodejs.org/job/node-test-pull-request/14266/

mcollina commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1361/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1362/

jasnell commented:
@nodejs/tsc ... any objections to this landing in 10.0.0?

jasnell commented:
Landed in f2d112c, bfe54df, and 2b0825e

mcollina commented:
+1 for me
